Refactor machine ownership to use player uuid

omergunr100 opened the pull request:
## What
Stores placer uuid instead of ownership object.
Ownership objects still exist but are generated in the getter.

## Outcome
This should solve some edge cases in our current implementation:
1) machines would cache teams that no longer exist.
2) when a player leaves the team the machine ownership object would keep both the team and player even though one should be invalid. (in this implementations machines belong to the player first and team second)

## Potential Compatibility Issues
Will remove all machine ownership data from existing worlds.
Moved ownership from block entity to meta machine so...
Any addon that re-assigned ownership will need to change from machine owner object to player uuid.

YoungOnionMC committed:
start of ownership syncing

YoungOnionMC committed:
im stuff

YoungOnionMC committed:
spotless

omergunr100 committed:
lombok and stop argonauts invited from accessing machines

omergunr100 committed:
cleanup

omergunr100 committed:
move ownership to placer uuid

omergunr100 committed:
missed one

YoungOnionMC commented on src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java:
```diff
@@ -140,6 +143,34 @@ public void setChanged() {
+            return null;
+        }
+        if (IMachineOwner.MachineOwnerType.FTB.isAvailable()) {
+            Optional<Team> team = FTBTeamsAPIImpl.INSTANCE.getManager().getTeamForPlayerID(uuid);
```
use the proper api calls here, see #2901

omergunr100 replied:
I thought your PR fixes it

omergunr100 committed:
Merge remote-tracking branch 'refs/remotes/official/yo/ownership-syncing' into oe/ownership-refactor

# Conflicts:
#	src/main/java/com/gregtechceu/gtceu/common/machine/owner/ArgonautsOwner.java
#	src/main/java/com/gregtechceu/gtceu/common/machine/owner/FTBOwner.java

omergunr100 committed:
fix missed spots and merge issues

omergunr100 committed:
consistency

krossgg commented on src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java:
```diff
@@ -140,6 +142,34 @@ public void setChanged() {
         }
     }
 
+    @SuppressWarnings({ "UnstableApiUsage", "deprecation" })
```
Mark `@Nullable`

krossgg commented on src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java:
```diff
@@ -140,6 +142,34 @@ public void setChanged() {
+        return new PlayerOwner(ownerUUID);
+    }
+
+    public PlayerOwner getPlayerOwner() {
```
Mark `@Nullable`

krossgg commented on src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java:
```diff
@@ -324,17 +354,19 @@ public AABB getRenderBoundingBox() {
+        if (tag.contains("ownerUUID")) {
+            ownerUUID = tag.getUUID("ownerUUID");
+        } else {
+            ownerUUID = null;
```
Else branch unnecessary, default initialization will make it null, then it will be set if the tag exists

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/data/GTSyncedFieldAccessors.java:
```diff
@@ -2,10 +2,7 @@
-import com.gregtechceu.gtceu.syncdata.GTRecipePayload;
-import com.gregtechceu.gtceu.syncdata.GTRecipeTypeAccessor;
-import com.gregtechceu.gtceu.syncdata.MaterialPayload;
+import com.gregtechceu.gtceu.syncdata.*;
```
why all the star imports :(

screret replied:
IMO it's fine

omergunr100 replied:
idk, it was from onions branch, I didn't touch this file

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/ArgonautsOwner.java:
```diff
@@ -38,28 +33,20 @@ public void save(CompoundTag tag) {
     @Override
     public void load(CompoundTag tag) {
         this.playerUUID = tag.getUUID("playerUUID");
-        this.server = ServerLifecycleHooks.getCurrentServer();
```
Is the server actually unnecessary here? Onion moment

krossgg reviewed (changes requested):
One further thought - maybe you could create a static UUID->IMachineOwner cache so that you don't have to create new instances of owners every time you call a `get`. Doesn't need to be persisted to savedata or anything.

screret commented on src/main/java/com/gregtechceu/gtceu/api/blockentity/MetaMachineBlockEntity.java:
```diff
@@ -140,6 +142,34 @@ public void setChanged() {
+            }
+        }
+        if (IMachineOwner.MachineOwnerType.ARGONAUTS.isAvailable()) {
+            var server = ServerLifecycleHooks.getCurrentServer();
```
you want to make this so it can be called on the client as well, so check if we're on the logical (not physical!) client and get the team from the client api instead.

screret commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/ArgonautsOwner.java:
```diff
@@ -38,28 +33,20 @@ public void save(CompoundTag tag) {
-        this.server = ServerLifecycleHooks.getCurrentServer();
-        var handler = GuildHandler.read(server);
-        this.guild = GuildHandler.API.get(server, tag.getUUID("guildUUID"));
+        this.guild = GuildClientApi.API.get(tag.getUUID("guildUUID"));
```
now this'll crash on a server instead :)

screret commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/ArgonautsOwner.java:
```diff
@@ -38,28 +33,20 @@ public void save(CompoundTag tag) {
-
-        return false;
+        var otherGuild = GuildClientApi.API.getPlayerGuild(player.getUUID());
+        return otherGuild != null && otherGuild.equals(this.guild);
```
ditto

omergunr100 commented:
tbh ownership should be moved from the block entity to the machine since the block entity holder can't be synced and there's no reason for the owner field to randomly just sit there.

omergunr100 committed:
move ownership from block entity to meta machine

omergunr100 committed:
make ownership objects safe for clients (only the parts that matter)

omergunr100 committed:
cleanup and prevent team/guild caching to prevent stale data when caching ownership objects

omergunr100 committed:
cache ownership objects

omergunr100 committed:
cache player owners

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/IMachineOwner.java:
```diff
@@ -24,7 +31,38 @@ public sealed interface IMachineOwner permits PlayerOwner, ArgonautsOwner, FTBOw
+            owner = new ArgonautsOwner(playerUUID);
+        } else {
+            owner = getPlayerOwner(playerUUID);
+        }
```
Stylistic choice, so up to you, but you could pull this out into a private method `createNewOwner`, then just do a oneline `return CACHE.computeIfAbsent(playerUUID, IMachineOwner::createNewOwner)` after the initial null check.

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/IMachineOwner.java:
```diff
@@ -24,7 +31,38 @@ public sealed interface IMachineOwner permits PlayerOwner, ArgonautsOwner, FTBOw
+        }
+        var owner = new PlayerOwner(playerUUID);
+        PLAYER_OWNERS.put(playerUUID, owner);
+        return owner;
```
Same here, can be replaced with `return PLAYER_OWNERS.computeIfAbsent(playerUUID, PlayerOwner::new);`

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/IMachineOwner.java:
```diff
@@ -70,7 +110,7 @@ static boolean canBreakOwnerMachine(Player player, IMachineBlockEntity machine)
 
```
While you're here, maybe a default equals/hashcode that just defers to equals/hashcode of `getUUID()` so that its possible to tell if two machines are owned by the same owner

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/FTBOwner.java:
```diff
@@ -1,93 +1,113 @@
+        if (FTBTeamsAPI.api().isManagerLoaded()) {
+            return FTBTeamsAPI.api().getManager().getPlayerTeamForPlayerID(playerUUID).orElse(null);
+        } else if (FTBTeamsAPI.api().isClientManagerLoaded()) {
+            return FTBTeamsAPI.api().getClientManager().getTeams().stream()
```
I see that the [`ClientTeamManager`](https://github.com/FTBTeam/FTB-Teams/blob/c864b003e95b6296f42bbd591bf10f945bd90565/common/src/main/java/dev/ftb/mods/ftbteams/api/client/ClientTeamManager.java#L58) has `selfTeam()`, do you think that would work rather than streaming through all the known client teams?

omergunr100 replied:
There's no guarantee that one of the players is the local player (addons could use this system in unexpected ways)

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/PlayerOwner.java:
```diff
@@ -52,20 +56,20 @@ public String getName() {
-                cache.get(playerUUID).ifPresent(value -> playerName[0] = value.getName());
+        final var playerName = UsernameCache.getLastKnownUsername(playerUUID);
+        String online;
+        if (GTCEu.isClientThread()) {
+            var connection = Minecraft.getInstance().getConnection();
+            if (connection != null) {
+                online = String.valueOf(connection.getOnlinePlayerIds().contains(playerUUID));
+            } else {
+                online = "Not Available";
             }
-            isOnline = false;
+        } else {
+            online = String
+                    .valueOf(ServerLifecycleHooks.getCurrentServer().getPlayerList().getPlayer(playerUUID) != null);
         }
-        compList.add(Component.translatable("behavior.portable_scanner.player_name", playerName[0], isOnline));
+        compList.add(Component.translatable("behavior.portable_scanner.player_name", playerName, online));
```
Maybe this could be extracted out into a separate method? I see that its duplicated in all 3 owner impls. Up to you.

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/PlayerOwner.java:
```diff
@@ -1,20 +1,24 @@
 import java.util.List;
 import java.util.UUID;
 
+@NoArgsConstructor
```
Consistency, add `@AllArgsConstructor`

krossgg reviewed (changes requested):
Another note, are the `IMachineOwner` methods `save`, `load`, and `create` (+ impls) still necessary? Seems that the owner itself is no longer serialized unless I missed something.

omergunr100 committed:
equals and hashcode

omergunr100 committed:
replace literals with lang entries.
extract common player information provider to interface.
remove save and load methods.

omergunr100 committed:
add uuid based permission methods.
changed internal method access to private.

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/ArgonautsOwner.java:
```diff
@@ -1,99 +1,92 @@
 import java.util.UUID;
 
+@SuppressWarnings({ "UnstableApiUsage", "removal", "deprecation" })
+@NoArgsConstructor
```
Last nitpick, I promise :^)

As far as I can tell, we no longer mutate owner objects, as we don't need to load them from tag. So, can we remove the noargs, make `playerUUID` final, and potentially even convert to a record class? Applicable to all 3 owner types of course.

If you want to go further, you could even make `IMachineOwner` into a abstract sealed class and move `playerUUID`and `hashcode/equals` upwards.

Just some thoughts.

omergunr100 replied:
Instead of locking down the machine ownership system I thought we should provide a way to add more options, at least for modpack makers who might want to change something about the default ownership system.

I clarified in the javadoc that it's not meant for addons.

krossgg commented on src/main/java/com/gregtechceu/gtceu/common/machine/owner/IMachineOwner.java:
```diff
@@ -1,67 +1,111 @@
+     * @param playerUUID the uuid of the player who owns the machine
+     * @return ownership object
+     */
+    @ApiStatus.Internal
```
Can remove this annotation and the docs since it's a private method now. Not accessible to anyone else. Maybe add doc comments to the `get(Player)Owner` methods

omergunr100 committed:
change machine owner from interface to abstract class and rename

omergunr100 committed:
refactor MachineOwner:
remove type enum.
move playerUUID to MachineOwner.
enable override of default ownership generator via an event. (made for mod-pack creators)

omergunr100 committed:
Merge branch '1.20.1' into oe/ownership-refactor

omergunr100 committed:
seal MachineOwner and unseal inheriting classes

omergunr100 committed:
make effectively final fields final

Ghostipedia committed:
Merge branch '1.20.1' into oe/ownership-refactor
